File: clrsPython/Lecture_1/Insertion_Sort.py
def Insertion_Sort(A):
    # A is the input array
    # Insertion_Sort function can permutate A as a nondecreasing array
    n = len(A) # length of A
    for i in range(1,n):
        key = A[i]
        j = i-1
        while j >= 0 and key < A[j]: # move key until key not less than A[j]
            A[j+1] = A[j] # let the jth element move backward
            j = j - 1
        A[j+1] = key
    return A

def Insertion_Sort_nonincreasing(A):
    # A is the input array
    # Insertion_Sort function can permutate A as a nonincreasing array
    n = len(A)  # length of A
    for i in range(1, n):
        key = A[i]
        j = i - 1
        while j>=0 and key >= A[j]: # move key until key not bigger than A[j]
            A[j+1] = A[j]
            j = j -1
        A[j+1] = key
    return A

def Insertion_Sort_reverse(A):
    # reverse way to implement Insertion_Sort
    # A is the input array
    # Insertion_Sort function can permutate A as a nondecreasing array
    n = len(A) # length of A
    j = n - 1
    while j >= 0:
        key = A[j]
        i = j + 1
        while i <= (n-1) and key >= A[i]: # move forward until not bigger than A[i]
            A[i-1] = A[i]
            i +=1
        A[i-1] = key
        j = j -1
    return A

File: clrsCpp/lecture_1/Insertion_Sort.cpp
#pragma once
#include <iostream>

int* Insertion_Sort(int* array, int length)
{
    // Use Insertion_sort to permutate the array as a nondecreasing array
    for (int i = 1; i < length; i++) // iteration from 2nd element in array
    {
        int key = *(array + i);
        int j = i - 1;
        while (j >= 0 && *(array + j) > key)
        {
            *(array + j + 1) = *(array + j); // move jth element to (j+1)th position
            j--;
        }
        *(array + j + 1) = key;
    }
    return array;
}

int* Insertion_Sort_nonincreasing(int* array, int length)
{
    // Use Insertion_sort to permutate the array as a nonincreasing array
    for (int i = 1; i < length; i++) // iteration from 2nd element in array
    {
        int key = *(array + i);
        int j = i - 1;
        while (j >= 0 && *(array + j) < key)
        {
            *(array + j + 1) = *(array + j); // move jth element to (j+1)th position
            j--;
        }
        *(array + j + 1) = key;
    }
    return array;
}



